cpp/src/parquet/arrow/arrow_reader_writer_test.cc

```diff
@@ -4138,6 +4139,74 @@ TEST_P(TestArrowWriteDictionary, Statistics) {
 INSTANTIATE_TEST_SUITE_P(WriteDictionary, TestArrowWriteDictionary,
                          ::testing::Values(ParquetDataPageVersion::V1,
                                            ParquetDataPageVersion::V2));
+
+TEST_P(TestArrowWriteDictionary, StatisticsUnifiedDictionary) {
+  // Two chunks, with a shared dictionary
+  std::shared_ptr<::arrow::Table> table;
+  std::shared_ptr<::arrow::DataType> dict_type =
+      ::arrow::dictionary(::arrow::int32(), ::arrow::utf8());
+  std::shared_ptr<::arrow::Schema> schema =
+      ::arrow::schema({::arrow::field("values", dict_type)});
+  {
+    // It's important there are no duplicate values in the dictionary, otherwise
+    // we trigger the WriteDense() code path which side-steps dictionary encoding.
+    std::shared_ptr<::arrow::Array> test_dictionary =
+        ArrayFromJSON(::arrow::utf8(), R"(["b", "c", "d", "a"])");
+    std::vector<std::shared_ptr<::arrow::Array>> test_indices = {
+        ArrayFromJSON(::arrow::int32(),
+                      R"([0, null, 3, 0, null, 3])"),  // ["b", null "a", "b", null, "a"]
+        ArrayFromJSON(
+            ::arrow::int32(),
+            R"([0, 3, null, 0, null, 1])")};  // ["b", "c", null, "b", "c", null]
```

Comment:
```suggestion
            R"([0, 1, null, 0, 1, null])")};  // ["b", "c", null, "b", "c", null]
```

I like what you have in the comment because then the min/max of row group 0 / chunk 0  is different from row group 0 / chunk 1.  Right now it looks like your indices don't match your comment and we have:

// ["b", null, "a", "b", null, "c"]

This leads to a/b being the min/max in stats0 but a/b is the min/max in both chunks of stats0.  To reproduce I think we want what you have in the comment which would mean chunk 0 is a/b and chunk 1 is b/c and so stats0 should be a/c.

Reply:
You are correct; I updated the data but forgot the comment. I did a weird thing where the chunks are 6/6 but the row groups are 9/3, anticipating that hit more interesting conditions in the writer (but maybe this is unnecessary).

Reply:
So I like the data as is. The first row group contains the first chunk plus the first three rows of the next chunk. Then the second group contains the last three elements of the second chunk. I've updated the comment so it is accurate again.

Reply:
Write...so I think (but could be wrong) this would lead to three calls to `WriteArrowDictionary`:

Call #1: (no previous dictionary) min=a, max=b, nulls=2
Call #2: (previous dictionary is equal) min=a, max=b, nulls=1
Call #3: (no previous dictionary) min=b, max=c, nulls=1

So if the bug was still in place, and it was using the first chunk to determine row-group statistics, it would still get the correct answer in this case.

Admittedly, the null count would still be wrong (it would report 2 nulls for stat0), so the test case itself wouldn't pass with the old code.  But I think it would get further than it should.

Reply:
Ah yes I see now. You are correct (just verified in lldb). I will change it so call 1 and 2 will have a different max.

(Also funny to realize how much PR 1, 2, and 3 of this repo have been mentioned 🤣 )